File: nodejs/refactoring-methods/01-extract-method.js
/**
 * 1. Method Extraction (Extract Method)
 *
 * BEFORE: A method contains too much logic, making it hard to understand
 */
class OrderProcessorBefore {
    processOrder(order) {
        // Validate order
        if (order.total <= 0) {
            throw new Error('Invalid order total');
        }

        // Calculate tax
        const tax = order.subtotal * 0.08;

        // Calculate shipping
        const shipping = order.weight > 10 ? 15.00 : 5.00;

        // Calculate total
        const total = order.subtotal + tax + shipping;

        // Save to database
        this.saveOrder(order, total);

        return total;
    }

    saveOrder(order, total) {
        // Database save logic
    }
}

/**
 * AFTER: Extract methods to separate concerns
 */
class OrderProcessorAfter {
    processOrder(order) {
        this.validateOrder(order);

        const tax = this.calculateTax(order);
        const shipping = this.calculateShipping(order);
        const total = this.calculateTotal(order, tax, shipping);

        this.saveOrder(order, total);

        return total;
    }

    validateOrder(order) {
        if (order.total <= 0) {
            throw new Error('Invalid order total');
        }
    }

    calculateTax(order) {
        return order.subtotal * 0.08;
    }

    calculateShipping(order) {
        return order.weight > 10 ? 15.00 : 5.00;
    }

    calculateTotal(order, tax, shipping) {
        return order.subtotal + tax + shipping;
    }

    saveOrder(order, total) {
        // Database save logic
    }
}

/**
 * 2. Embedding a method (Inline Method)
 *
 * BEFORE: A method is too simple and adds no value
 */
class UserBefore {
    getFullName() {
        return this.getFirstName() + ' ' + this.getLastName();
    }

    getFirstName() {
        return this.firstName;
    }

    getLastName() {
        return this.lastName;
    }
}

/**
 * AFTER: Inline the simple method
 */
class UserAfter {
    getFullName() {
        return this.firstName + ' ' + this.lastName;
    }

    // getFirstName() and getLastName() methods removed
}

module.exports = {
    OrderProcessorBefore,
    OrderProcessorAfter,
    UserBefore,
    UserAfter
};

File: python/refactoring-methods/01-extract-method.py
# 1. Method Extraction (Extract Method)
#
# BEFORE: A method contains too much logic, making it hard to understand
class OrderProcessorBefore:
    def process_order(self, order):
        # Validate order
        if order['total'] <= 0:
            raise ValueError('Invalid order total')

        # Calculate tax
        tax = order['subtotal'] * 0.08

        # Calculate shipping
        shipping = 15.00 if order['weight'] > 10 else 5.00

        # Calculate total
        total = order['subtotal'] + tax + shipping

        # Save to database
        self.save_order(order, total)

        return total

    def save_order(self, order, total):
        # Database save logic
        pass

# AFTER: Extract methods to separate concerns
class OrderProcessorAfter:
    def process_order(self, order):
        self.validate_order(order)

        tax = self.calculate_tax(order)
        shipping = self.calculate_shipping(order)
        total = self.calculate_total(order, tax, shipping)

        self.save_order(order, total)

        return total

    def validate_order(self, order):
        if order['total'] <= 0:
            raise ValueError('Invalid order total')

    def calculate_tax(self, order):
        return order['subtotal'] * 0.08

    def calculate_shipping(self, order):
        return 15.00 if order['weight'] > 10 else 5.00

    def calculate_total(self, order, tax, shipping):
        return order['subtotal'] + tax + shipping

    def save_order(self, order, total):
        # Database save logic
        pass

# 2. Embedding a method (Inline Method)
#
# BEFORE: A method is too simple and adds no value
class UserBefore:
    def __init__(self, first_name, last_name):
        self.first_name = first_name
        self.last_name = last_name

    def get_full_name(self):
        return self.get_first_name() + ' ' + self.get_last_name()

    def get_first_name(self):
        return self.first_name

    def get_last_name(self):
        return self.last_name

# AFTER: Inline the simple method
class UserAfter:
    def __init__(self, first_name, last_name):
        self.first_name = first_name
        self.last_name = last_name

    def get_full_name(self):
        return self.first_name + ' ' + self.last_name

    # get_first_name() and get_last_name() methods removed
